### PR TITLE
Block module: stage 1

### PR DESCRIPTION
**Pull request by alex-arugon**
One should know - the module is not ready.

### CODE REVIEW

File: api/service/block.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    public list(filter: IListParams): Response<{ blocks: Block[], count: number }> {
+        const params: IListParams = {};
+        const where: string[] = [];
+
+        if (filter.generatorPublicKey) {
+            where.push('"b_generatorPublicKey"::bytea = ${generatorPublicKey}');
+            params.generatorPublicKey = filter.generatorPublicKey;
+        }
+
+        // FIXME: Useless condition
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    private blockRepo = new BlockRepo();
+    private lastBlock: Block;
+    private sortFields = [
+        'id',
+        'timestamp',
+        'height',
+        'previousBlock',
+        'totalAmount',
+        'totalFee',
+        'reward',
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+export class BlockService {
+    private blockRepo = new BlockRepo();
+    private lastBlock: Block;
+    private sortFields = [
+        'id',
+        'timestamp',
+        'height',
+        'previousBlock',
+        'totalAmount',
+        'totalFee',
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+export class BlockService {
+    private blockRepo = new BlockRepo();
+    private lastBlock: Block;
+    private sortFields = [
+        'id',
+        'timestamp',
+        'height',
+        'previousBlock',
+        'totalAmount',
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        'totalFee',
+        'reward',
+        'numberOfTransactions',
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        'generatorPublicKey'
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+export class BlockService {
+    private blockRepo = new BlockRepo();
+    private lastBlock: Block;
+    private sortFields = [
+        'id',
+        'timestamp',
+        'height',
+        'previousBlock',
```
Comment (Verbalist) [inline api/service/block.ts]: useless to

File: api/service/block.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+            params.previousBlock = filter.previousBlock;
+        }
+
+        if (filter.height === 0 || filter.height > 0) {
+            where.push('"b_height" = ${height}');
+            params.height = filter.height;
+        }
+
+        // FIXME: Useless condition
+        if (filter.totalAmount >= 0) {
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+            where.push('"b_totalAmount" = ${totalAmount}');
+            params.totalAmount = filter.totalAmount;
+        }
+
+        // FIXME: Useless condition
+        if (filter.totalFee >= 0) {
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: api/service/block.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+            where.push('"b_totalFee" = ${totalFee}');
+            params.totalFee = filter.totalFee;
+        }
+
+        // FIXME: Useless condition
+        if (filter.reward >= 0) {
```
Comment (Verbalist) [inline api/service/block.ts]: remove

File: core/controller/block.ts [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+        }
+    }
+
+    @ON('BLOCKCHIN_READY')
+    public loadLastNBlocks(): void {
+        const blocks: string[] = this.blockRepo.loadLastNBlocks();
+        this.blockService.setLastNBlocks(blocks);
+    }
+
+    @ON('NEW_BLOCK')
```
Comment (Verbalist) [inline core/controller/block.ts]: NEW_BLOCKS

File: core/service/block.ts [hunk 12]
```diff
@@ -26,5 +71,683 @@ export class BlockService {
...
+        this.saveBlock(block); // config.genesis.block)
+    }
+
+    public saveBlock(block: Block): void {
+        this.dbSave(block);
+        this.promiseTransactions(block);
+        this.afterSave(block);
+    }
+
+    private dbSave(block: Block): IDBBlockSave {
```
Comment (Verbalist) [inline core/service/block.ts]: It`s a repo method

File: core/service/block.ts [hunk 13]
```diff
@@ -26,5 +71,683 @@ export class BlockService {
...
+                payloadLength: block.payloadLength,
+                payloadHash,
+                generatorPublicKey,
+                blockSignature
+            }
+        };
+    }
+
+    private afterSave(block: Block): void {
+        // call 'transactionsSaved' on bus
```
Comment (Verbalist) [inline core/service/block.ts]: yes, only bus call

File: core/service/block.ts [hunk 14]
```diff
@@ -26,5 +71,683 @@ export class BlockService {
...
+        // ?
+        block.transactions.forEach((transaction) => library.logic.transaction.afterSave(transaction));
+    }
+
+    private async promiseTransactions(block: Block): Promise<object> {
+        return await library.logic.transaction.dbSave(transaction);
+    }
+
+    public applyGenesisBlock(block: Block): void {
```
Comment (Verbalist) [inline core/service/block.ts]: use new apply genesis block function

File: core/service/block.ts [hunk 15]
```diff
@@ -26,5 +71,683 @@ export class BlockService {
...
+            const sender = modules.accounts.setAccountAndGet({ publicKey: transaction.senderPublicKey });
+            this.applyTransaction(block, transaction, sender);
+            tracker.applyNext();
+        });
+        this.setLastBlock(block);
+        // busrx call to ROUNDS_TICK
+        modules.rounds.tick(block, cb);
+    }
+
+    public getBlockProgressLogger(transactionsCount: number, logsFrequency: number, msg: string) {
```
Comment (Verbalist) [inline core/service/block.ts]: useless

File: core/service/block.ts [hunk 16]
```diff
@@ -26,5 +71,683 @@ export class BlockService {
...
+            this.log = function () {
+                library.logger.info(msgBlock, `${((this.applied / this.target) * 100)
+                    .toPrecision(4)} % : applied ${this.applied} of ${this.target} transactions`);
+            };
+        }
+
+        return new BlockProgressLogger(transactionsCount, logsFrequency, msg);
+    }
+
+    private applyTransaction(block: Block, transaction: Transaction<object>, sender): void {
```
Comment (Verbalist) [inline core/service/block.ts]: check new implementation in master and rewrite it

File: api/controller/block.ts [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        return new Response({ data: { epoch: constants.epochTime }});
+    }
+
+    @GET('/height') // /getHeight
+    public getHeight(): Response<{ height: number }> {
+        const height = 0; // modules.blocks.lastBlock.get().height
+        return new Response({ data: { height }});
+    }
+
+    @GET('/fee') // /getFee
```
[the remark this answers is not in the document]
Comment (alex-arugon) [reply api/controller/block.ts]: called from frontend, e.g. freezeAmountController

File: api/controller/block.ts [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+
+    @GET('/reward') // /getReward
+    public getReward(): Response<{ reward: number }> {
+        // __private.blockReward.calcReward(modules.blocks.lastBlock.get().height)}
+        return new Response({ data: { reward: 0 }});
+    }
+
+    @GET('/supply') // /getSupply
+    public getSupply(): Response<{ supply: number }> {
+        return new Response({ data: { supply: this.blockReward.calcSupply(this.blockService.getLastBlock().height) }});
```
[the remark this answers is not in the document]
Comment (alex-arugon) [reply api/controller/block.ts]: please, suggest changes

File: api/controller/block.ts [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+const constants = require('../../helpers/constants.js');
+const schema = require('../../schema/blocks.js');
+
+@Controller('/blocks')
+export class BlockController {
+    private blockReward = new BlockReward();
+    private blockService = new BlockService();
+    private blockRepo = new BlockRepo();
+
+    constructor() {}
```
Comment (macaqueee) [inline api/controller/block.ts]: Initialization of components place here, in constructor.

Comment (axelsheva) [reply api/controller/block.ts]: Not fixed

File: api/controller/system.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
+import Response from 'shared/model/response';
+import {BlockService} from 'api/service/block';
+
+const constants = require('../../backlog/helpers/constants');
+
+class RoundService {}
+
+@Controller('?')
```
Comment (axelsheva) [inline api/controller/system.ts]: Is it correct?

File: api/service/block.ts [hunk 21]
```diff
@@ -0,0 +1,112 @@
...
+        'timestamp',
+        'height'
+    ];
+
+    public loadBlocksData(params: ILoadBlockParams) : Block[] {
+        return [];
+    }
+
+    public getLastBlock(): Block {
+        // load block from db?
```
Comment (axelsheva) [inline api/service/block.ts]: Last block must be stored in memory

File: api/service/block.ts [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+            where.push('"b_generatorPublicKey"::bytea = ${generatorPublicKey}');
+            params.generatorPublicKey = filter.generatorPublicKey;
+        }
+
+        if (filter.previousBlock) {
+            where.push('"b_previousBlock" = ${previousBlock}');
+            params.previousBlock = filter.previousBlock;
+        }
+
+        if (filter.height === 0 || filter.height > 0) {
```
Comment (axelsheva) [inline api/service/block.ts]: `filter.height === 0` can block height be zero?

File: core/controller/block.ts [hunk 23]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @ON('NEW_BLOCKS')
+    public updateLastNBlocks(block: Block): Response<void> {
+        this.blockService.updateLastNBlocks(block);
+        return new Response<void>();
+    }
+
+    @RPC('GET_COMMON_BLOCK')
+    // called from UI
```
Comment (axelsheva) [inline core/controller/block.ts]: This logic should not be called from RPC

File: api/controller/system.ts (hunk 20 above)
Comment (alex-arugon) [reply api/controller/system.ts]: No it isn't, but the only thing I know about this handler is the action it should perform.

File: api/controller/block.ts (hunk 19 above)
Comment (alex-arugon) [reply api/controller/block.ts]: Will be fixed while moving all related services to singletons.

File: api/service/block.ts (hunk 21 above)
Comment (alex-arugon) [reply api/service/block.ts]: Ok, but the thing is API doesn't know about last block changing. So currently nothing sets last block to memory, so nothing is returned by this function.

File: api/service/block.ts (hunk 22 above)
Comment (alex-arugon) [reply api/service/block.ts]: I hope it can't, fixing

File: core/controller/block.ts (hunk 23 above)
Comment (alex-arugon) [reply core/controller/block.ts]: Also used for sync between nodes, but called from current UI as well. I'm not sure if this method is necessary here.

Comment (axelsheva) [reply core/controller/block.ts]: Method for sync between nodes cannot be RPC or API. It must be private

Comment (axelsheva) [reply core/controller/block.ts]: But here can be a method for getting a common block from another nodes. Like a https://github.com/AraiEzzra/DDKCORE/blob/ed086aad0cb545fd952f5ef1da38820c89014200/src/modules/transport.js#L556

**Pull request merged**